### PR TITLE
[JUJU-4125] Replace usages of external controller Save to new domain

### PR DESCRIPTION
**Pull request by nvinuesa**
Replace the Save method from the legacy state to the new UpdateExternalController on new domain using DQlite. Inject the external controller service (from domain) into the facades that now depend on it and fix corresponding tests.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```
go test github.com/juju/juju/apiserver/facades/client/application/... -gocheck.v
go test github.com/juju/juju/apiserver/facades/client/controller/... -gocheck.v
go test github.com/juju/juju/apiserver/facades/controller/migrationtarget/... -gocheck.v
go test github.com/juju/juju/apiserver/facades/controller/remoterelations/... -gocheck.v
go test github.com/juju/juju/apiserver/common/... -gocheck.v
```

### CODE REVIEW

File: apiserver/common/controllerconfig.go [hunk 1]
```diff
@@ -101,12 +117,13 @@ func (s *ControllerConfigAPI) getModelControllerInfo(model params.Entity) (param
 	}
 
 	logger.Debugf("found migrated model on another controller, saving the information")
-	_, err = ec.Save(crossmodel.ControllerInfo{
+	err = s.ecService.UpdateExternalController(ctx, crossmodel.ControllerInfo{
```
Comment (nvinuesa) [inline apiserver/common/controllerconfig.go:120]: This is (one of the examples of)  the main change on this patch.

File: apiserver/facades/client/application/application.go [hunk 2]
```diff
@@ -1944,12 +1966,13 @@ func (api *APIBase) consumeOne(arg params.ConsumeApplicationArg) error {
 		// a different controller.
 		if controllerTag.Id() != api.backend.ControllerTag().Id() {
 			externalControllerUUID = controllerTag.Id()
-			if _, err = api.backend.SaveController(crossmodel.ControllerInfo{
+			if err = api.ecService.UpdateExternalController(ctx, crossmodel.ControllerInfo{
```
Comment (nvinuesa) [inline apiserver/facades/client/application/application.go:1969]: Another example, main change.

File: apiserver/facades/controller/migrationtarget/migrationtarget.go [hunk 3]
```diff
@@ -200,18 +208,18 @@ func (api *API) Activate(args params.ActivateModelArgs) error {
...
 	// model relations to the source controller that were local but
 	// now need to be external after migration.
-	ec := api.state.NewExternalControllers()
 	if len(args.CrossModelUUIDs) > 0 {
 		cTag, err := names.ParseControllerTag(args.ControllerTag)
 		if err != nil {
 			return errors.Trace(err)
 		}
-		_, err = ec.Save(crossmodel.ControllerInfo{
+		err = api.ecService.UpdateExternalController(ctx, crossmodel.ControllerInfo{
```
Comment (nvinuesa) [inline apiserver/facades/controller/migrationtarget/migrationtarget.go:216]: Another example of the main change on this patch.

File: apiserver/facades/controller/migrationtarget/migrationtarget.go [hunk 4]
```diff
@@ -225,12 +233,12 @@ func (api *API) Activate(args params.ActivateModelArgs) error {
 	}
 	for _, app := range remoteApps {
 		var sourceControllerUUID string
-		extInfo, err := ec.ControllerForModel(app.SourceModel().Id())
+		extInfo, err := api.ecService.ControllerForModel(ctx, app.SourceModel().Id())
```
Comment (nvinuesa) [inline apiserver/facades/controller/migrationtarget/migrationtarget.go:236]: Another one.

**Review by nvinuesa — COMMENTED** [review]
This patch looks big but there is only one main set of changes (replacing the legacy state with domain calls), and adding context to some of the facade functions to be passed to the domain calls. This changes along with the new mocks impact some tests.

File: apiserver/facades/controller/migrationtarget/migrationtarget_test.go [hunk 5]
```diff
@@ -51,6 +52,9 @@ type Suite struct {
 var _ = gc.Suite(&Suite{})
 
 func (s *Suite) SetUpTest(c *gc.C) {
+	ctrl := gomock.NewController(c)
```
Comment (manadart) [inline apiserver/facades/controller/migrationtarget/migrationtarget_test.go:55]: Use a `setupMocks` method as mentioned prior.

File: apiserver/facades/client/application/application.go [hunk 6]
```diff
@@ -58,6 +62,12 @@ var ClassifyDetachedStorage = storagecommon.ClassifyDetachedStorage
 
 var logger = loggo.GetLogger("juju.apiserver.application")
 
+type EcService interface {
```
Comment (manadart) [inline apiserver/facades/client/application/application.go:65]: Convention would tend to have this as `ECService`.

File: apiserver/facades/agent/secretsmanager/register.go [hunk 7]
```diff
@@ -74,9 +78,18 @@ func NewSecretManagerAPI(context facade.Context) (*SecretsManagerAPI, error) {
...
+			ecstate.NewState(domain.NewTxnRunnerFactory(context.ControllerDB)),
+			domain.NewWatcherFactory(
+				context.ControllerDB,
+				context.Logger().Child("secretsmanager"),
+			),
+		),
+	)
 	remoteClientGetter := func(uri *coresecrets.URI) (CrossModelSecretsClient, error) {
-		info, err := controllerAPI.ControllerAPIInfoForModels(params.Entities{Entities: []params.Entity{{
+		info, err := controllerAPI.ControllerAPIInfoForModels(stdContext.Background(), params.Entities{Entities: []params.Entity{{
```
Comment (manadart) [inline apiserver/facades/agent/secretsmanager/register.go:92]: `stdContext.TODO()`

**Review by manadart — COMMENTED** [review]
Mostly good; couple of changes.

File: apiserver/facades/controller/migrationtarget/migrationtarget_test.go (hunk 5 above)
Comment (nvinuesa) [reply apiserver/facades/controller/migrationtarget/migrationtarget_test.go:55]: yes of course, done

File: apiserver/facades/client/application/application.go (hunk 6 above)
Comment (nvinuesa) [reply apiserver/facades/client/application/application.go]: sure, fixed

File: apiserver/facades/agent/secretsmanager/register.go (hunk 7 above)
Comment (nvinuesa) [reply apiserver/facades/agent/secretsmanager/register.go]: done

File: apiserver/facades/controller/migrationtarget/migrationtarget_test.go (hunk 5 above)
Comment (nvinuesa) [reply apiserver/facades/controller/migrationtarget/migrationtarget_test.go:55]: done

File: apiserver/facades/client/application/application.go (hunk 6 above)
Comment (nvinuesa) [reply apiserver/facades/client/application/application.go]: sure, fixed

File: apiserver/facades/agent/secretsmanager/register.go (hunk 7 above)
Comment (nvinuesa) [reply apiserver/facades/agent/secretsmanager/register.go]: done

**Review by manadart — APPROVED** [review]
This looks OK, but you will need a proper QA at some point, establishing and migrating a cross-model (cross-controller) relation and seeing that it is preserved.

File: apiserver/facades/client/application/backend.go [hunk 8]
```diff
@@ -250,9 +250,10 @@ type modelShim struct {
 
 type ExternalController state.ExternalController
 
+// TODO(nvinuesa): remove this function when we cleanup externalcontroller
+// legacy state.
 func (s stateShim) SaveController(controllerInfo crossmodel.ControllerInfo, modelUUID string) (ExternalController, error) {
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply apiserver/facades/client/application/backend.go]: yes of course, done

File: apiserver/common/controllerconfig_test.go [hunk 9]
```diff
@@ -139,17 +150,21 @@ func (s *controllerInfoSuite) TestControllerInfoLocalModel(c *gc.C) {
...
 		Addrs:         []string{"192.168.1.1:12345"},
 		CACert:        testing.CACert,
 	}
-	_, err := ec.Save(info, modelUUID)
-	c.Assert(err, jc.ErrorIsNil)
-	cc := common.NewStateControllerConfig(s.localState)
-	results, err := cc.ControllerAPIInfoForModels(params.Entities{
+
+	s.ecService.EXPECT().ControllerForModel(gomock.Any(), modelUUID).
+		Times(1).
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply apiserver/common/controllerconfig_test.go]: done

File: apiserver/common/controllerconfig.go [hunk 10]
```diff
@@ -12,16 +14,31 @@ import (
...
 // facades - eg Provisioner and ControllerConfig.
 type ControllerConfigAPI struct {
-	st ControllerConfigState
+	ecService ECService
+	st        ControllerConfigState
 }
 
 // NewStateControllerConfig returns a new NewControllerConfigAPI.
-func NewStateControllerConfig(st ControllerConfigState) *ControllerConfigAPI {
+func NewStateControllerConfig(
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply apiserver/common/controllerconfig.go]: done

File: apiserver/facades/controller/migrationtarget/migrationtarget.go [hunk 11]
```diff
@@ -27,10 +27,21 @@ import (
 	"github.com/juju/juju/state/stateenvirons"
 )
 
+type EcService interface {
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply apiserver/facades/controller/migrationtarget/migrationtarget.go]: done

**Pull request merged**